picoquic/frames.c

```diff
@@ -940,108 +1020,44 @@ static picoquic_packet * picoquic_update_rtt(picoquic_cnx_t * cnx, uint64_t larg
+
+	picoquic_sack_item_t const * target_sack = &cnx->first_sack_item;
+	while (target_sack->next_sack != NULL)
 	{
-		ret = -1;
+		target_sack = target_sack->next_sack;
 	}
-	else
-	{
-		/* Find the oldest ACK range, in order to calibrate the
-		* extension of the largest number to 64 bits */
-
-		while (sack->next_sack != NULL)
-		{
-			sack = sack->next_sack;
-		}
-		target_sequence = sack->start_of_sack_range - 1;
+	uint64_t target_sequence = target_sack->start_of_sack_range - 1;
 
-		if (has_num_block)
-		{
-            num_block = bytes[byte_index++];
-		}
-        num_ts = bytes[byte_index++];
+	ret = picoquic_parse_ack_header(bytes, bytes_max, target_sequence,
+									&num_block, &num_ts, &largest,
+									&ack_delay, &mm, consumed);
 
-		/* decoding the largest */
-		switch (ll)
-		{
-		case 0:
-			largest = bytes[byte_index++];
-			largest = picoquic_get_packet_number64(target_sequence,
-				0xFFFFFFFFFFFFFF00ull, (uint32_t)largest);
-			break;
-		case 1:
-			largest = PICOPARSE_16(bytes + byte_index);
-			largest = picoquic_get_packet_number64(target_sequence,
-				0xFFFFFFFFFFFF0000ull, (uint32_t)largest);
-			byte_index += 2;
-			break;
-		case 2:
-			largest = PICOPARSE_32(bytes + byte_index);
-			largest = picoquic_get_packet_number64(target_sequence,
-				0xFFFFFFFF00000000ull, (uint32_t)largest);
-			byte_index += 4;
-			break;
-		case 3:
-			largest = PICOPARSE_64(bytes + byte_index);
-			byte_index += 8;
-			break;
-                default:
-                        DBG_FATAL_PRINTF("Internal error: invalid ll=%u", ll);
-                        break;
-		}
-		/* Skip everything else -- we are only interested in ACK parsing */
-
-        byte_index += 2; /* ACK delay */
-
-        /* last range and blocks */
-        switch (mm)
-        {
-        case 0:
-            byte_index += 1 + num_block*(1 + 1);
-            break;
-        case 1:
-            byte_index += 2 + num_block*(1 + 2);
-            break;
-        case 2:
-            byte_index += 4 + num_block*(1 + 4);
-            break;
-        case 3:
-            byte_index += 8 + num_block*(1 + 8);
-            break;
-        default:
-            DBG_FATAL_PRINTF("Internal error: invalid mm=%u", mm);
-            break;
-        }
+	if (ret == 0)
+	{
+        /* consume last range and blocks */
+		*consumed += (1 << mm) + num_block*(1 + (1 << mm));
 
         if (num_ts > 0)
         {
-            byte_index += 2 + num_ts * 3;
+            *consumed += 2 + num_ts * 3;
         }
 
-        *consumed = byte_index;
-
         /* Process the parsed ACK */
-		sack = &cnx->first_sack_item;
-		previous_sack = NULL;
-		sack_count = 0;
+		picoquic_sack_item_t * sack          = &cnx->first_sack_item;
+		picoquic_sack_item_t * previous_sack = NULL;
+		unsigned               sack_count    = 0;
 
 		do {
+			/* TODO: Clarify logic here and simplify */
 			if (sack->start_of_sack_range < largest &&
 				sack->end_of_sack_range < largest)
 			{
```

Comment:
I think you changed the logic there. Largest is the "largest seen". It does not mean that all packets before that were acked. The acknowledged ranges are "before largest". An ACK is acknowledged if it is present in one of these ranges, not just if it's sequence number is below largest.

Reply:
Are you referring to this the code below this comment?  I specifically did NOT change that code -- it is the original.  I added the comment, *because* it seemed wrong.

Reply:
You are right, I should have looked at the whole set.